In [ ]:
# import modules & set up logging
import gensim, logging, numpy as np
import help_functions as hf
import nltk
nltk.download('wordnet')

#@author: The first version of this code is the courtesy of Vadim Selyanik


logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)
lemmatizer = nltk.WordNetLemmatizer() # create a lemmatizer

sentences = []
file = open("lemmatized.text", "r")

for line in file: # read the file and create list which contains all sentences found in the text
    sentences.append(line.split())
# train word2vec on the two sentences


In [8]:
def runSynonymModel(dimension):
    threshold = 0.00055 # parameter for Word2vec


    sum = 0.0

    model = gensim.models.Word2Vec(sentences, min_count=1, sample=threshold, sg=1,size=dimension) # create model using Word2Ve with the given parameters
    #
    print(len(model.wv.vocab)) # check the length of the vocabulary which was formed by Word2Vec
    #The rest implements passing TOEFL tests
    i = 0 #counter for TOEFL tests
    number_of_tests = 80
    text_file = open('new_toefl.txt', 'r')
    right_answers = 0 # variable for correct answers
    number_skipped_tests = 0 # some tests could be skipped if there are no corresponding words in the vocabulary extracted from the training corpus
    while i < number_of_tests:
                line = text_file.readline() #read line in the file
                words = line.split() # extract words from the line
                try:
                    words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                             words] # lemmatize words in the current test
                    vectors = []
                    if words[0] in model: # check if there embedding for the query word
                        k = 1 #counter for loop iterating over 5 words in the test
                        vectors.append(model[words[0]])
                        while k < 5:
                            if words[k] in model: # if alternative has the embedding
                                vectors.append(model[words[k]]) #assing the learned vector
                            else: 
                                vectors.append(np.random.randn(dimension)) #assing random vector
                            k += 1
                        right_answers += hf.get_answer_mod(vectors) #find the closest vector and check if it is the correct answer

                except KeyError: # if there is no representation for the query vector than skip
                    number_skipped_tests += 1
                    print("skipped test: " + str(i) + "; Line: " + str(words))
                except IndexError:
                    print(i)
                    print(line)
                    print(words)
                    break
                i += 1
    text_file.close()
    sum += 100 * float(right_answers) / float(number_of_tests) #get the percentage
    print("Threshold ferq = "+ str(threshold)+" Percentage of correct answers: " + str(sum) + "%")
    return sum

In [10]:
dim10 = 0
dim100 = 0
dim1000 = 0
runs = 3
for i in range(runs):
    dim10 += runSynonymModel(10)
    dim100 += runSynonymModel(100)
    dim1000 += runSynonymModel(1000)
    
print("Dimension 10 mean accuracy = ",dim10/runs)
print("Dimension 100 mean accuracy = ",dim100/runs)
print("Dimension 1000 mean accuracy = ",dim1000/runs)

2021-02-23 11:36:21,927 : WARNING : consider setting layer size to a multiple of 4 for greater performance
2021-02-23 11:36:21,928 : INFO : collecting all words and their counts
2021-02-23 11:36:21,929 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2021-02-23 11:36:21,949 : INFO : PROGRESS: at sentence #10000, processed 112555 words, keeping 6354 word types
2021-02-23 11:36:21,976 : INFO : PROGRESS: at sentence #20000, processed 254983 words, keeping 10821 word types
2021-02-23 11:36:21,998 : INFO : PROGRESS: at sentence #30000, processed 367772 words, keeping 12635 word types
2021-02-23 11:36:22,033 : INFO : PROGRESS: at sentence #40000, processed 566573 words, keeping 17685 word types
2021-02-23 11:36:22,060 : INFO : PROGRESS: at sentence #50000, processed 724294 words, keeping 20496 word types
2021-02-23 11:36:22,086 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:36:22,111 : INFO : PROGRESS: at senten

2021-02-23 11:36:23,911 : INFO : PROGRESS: at sentence #710000, processed 10223733 words, keeping 68816 word types
2021-02-23 11:36:23,940 : INFO : PROGRESS: at sentence #720000, processed 10360488 words, keeping 69381 word types
2021-02-23 11:36:23,961 : INFO : PROGRESS: at sentence #730000, processed 10479153 words, keeping 69774 word types
2021-02-23 11:36:23,989 : INFO : PROGRESS: at sentence #740000, processed 10597851 words, keeping 70112 word types
2021-02-23 11:36:24,016 : INFO : PROGRESS: at sentence #750000, processed 10734438 words, keeping 70523 word types
2021-02-23 11:36:24,024 : INFO : collected 70583 word types from a corpus of 10766283 raw words and 752149 sentences
2021-02-23 11:36:24,027 : INFO : Loading a fresh vocabulary
2021-02-23 11:36:24,172 : INFO : effective_min_count=1 retains 70583 unique words (100% of original 70583, drops 0)
2021-02-23 11:36:24,174 : INFO : effective_min_count=1 leaves 10766283 word corpus (100% of original 10766283, drops 0)
2021-02-23 1

2021-02-23 11:37:27,273 : INFO : EPOCH 5 - PROGRESS: at 18.94% examples, 665337 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:28,275 : INFO : EPOCH 5 - PROGRESS: at 28.20% examples, 659682 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:29,288 : INFO : EPOCH 5 - PROGRESS: at 36.45% examples, 639529 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:30,297 : INFO : EPOCH 5 - PROGRESS: at 44.51% examples, 639531 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:31,310 : INFO : EPOCH 5 - PROGRESS: at 53.67% examples, 641321 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:32,320 : INFO : EPOCH 5 - PROGRESS: at 61.72% examples, 640822 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:33,322 : INFO : EPOCH 5 - PROGRESS: at 69.79% examples, 634588 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:34,324 : INFO : EPOCH 5 - PROGRESS: at 74.87% examples, 612103 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:37:35,327 : INFO : EPOCH 5 - PROGRESS: at 82.76% examples, 607216 words/s, in_qsiz

70583
Threshold ferq = 0.00055 Percentage of correct answers: 53.75%


2021-02-23 11:37:37,657 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:37:37,682 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:37:37,703 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:37:37,729 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:37:37,752 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:37:37,778 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:37:37,808 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keeping 30297 word types
2021-02-23 11:37:37,832 : INFO : PROGRESS: at sentence #150000, processed 2044830 words, keeping 31192 word types
2021-02-23 11:37:37,854 : INFO : PROGRESS: at sentence #160000, processed 2160734 words, k

2021-02-23 11:37:40,511 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:37:40,512 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:37:40,737 : INFO : estimated required memory for 70583 words and 100 dimensions: 91757900 bytes
2021-02-23 11:37:40,738 : INFO : resetting layer weights
2021-02-23 11:37:53,366 : INFO : training model with 3 workers on 70583 vocabulary and 100 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:37:54,378 : INFO : EPOCH 1 - PROGRESS: at 6.78% examples, 500364 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:55,383 : INFO : EPOCH 1 - PROGRESS: at 13.76% examples, 484561 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:56,392 : INFO : EPOCH 1 - PROGRESS: at 20.99% examples, 482850 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:57,401 : INFO : EPOCH 1 - PROGRESS: at 27.49% examples, 480647 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:37:58,405 : INFO : EPOCH 1

2021-02-23 11:38:54,009 : INFO : EPOCH 4 - PROGRESS: at 83.82% examples, 511733 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:38:55,073 : INFO : EPOCH 4 - PROGRESS: at 88.69% examples, 499954 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:38:56,104 : INFO : EPOCH 4 - PROGRESS: at 94.01% examples, 488426 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:38:57,130 : INFO : EPOCH 4 - PROGRESS: at 98.92% examples, 476963 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:38:57,303 : INFO : worker thread finished; awaiting finish of 2 more threads
2021-02-23 11:38:57,317 : INFO : worker thread finished; awaiting finish of 1 more threads
2021-02-23 11:38:57,322 : INFO : worker thread finished; awaiting finish of 0 more threads
2021-02-23 11:38:57,323 : INFO : EPOCH - 4 : training on 10766283 raw words (7341475 effective words) took 15.4s, 475697 effective words/s
2021-02-23 11:38:58,336 : INFO : EPOCH 5 - PROGRESS: at 5.94% examples, 432886 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:38:59,349 

70583
Threshold ferq = 0.00055 Percentage of correct answers: 71.25%


2021-02-23 11:39:14,492 : INFO : PROGRESS: at sentence #50000, processed 724294 words, keeping 20496 word types
2021-02-23 11:39:14,521 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:39:14,545 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:39:14,569 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:39:14,592 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:39:14,610 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:39:14,633 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:39:14,653 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:39:14,676 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keepin

2021-02-23 11:39:16,407 : INFO : Loading a fresh vocabulary
2021-02-23 11:39:16,506 : INFO : effective_min_count=1 retains 70583 unique words (100% of original 70583, drops 0)
2021-02-23 11:39:16,506 : INFO : effective_min_count=1 leaves 10766283 word corpus (100% of original 10766283, drops 0)
2021-02-23 11:39:16,711 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:39:16,715 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:39:16,717 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:39:16,943 : INFO : estimated required memory for 70583 words and 1000 dimensions: 599955500 bytes
2021-02-23 11:39:16,944 : INFO : resetting layer weights
2021-02-23 11:39:31,635 : INFO : training model with 3 workers on 70583 vocabulary and 1000 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:39:32,664 : INFO : EPOCH 1 - PROGRESS: at 1.40% examples, 79750 words/s, in_qsize 5, out_qsize 

2021-02-23 11:40:41,686 : INFO : EPOCH 1 - PROGRESS: at 85.06% examples, 89621 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:42,702 : INFO : EPOCH 1 - PROGRESS: at 86.13% examples, 89777 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:43,728 : INFO : EPOCH 1 - PROGRESS: at 87.18% examples, 89836 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:44,763 : INFO : EPOCH 1 - PROGRESS: at 88.47% examples, 89872 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:45,904 : INFO : EPOCH 1 - PROGRESS: at 90.08% examples, 89957 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:46,905 : INFO : EPOCH 1 - PROGRESS: at 91.37% examples, 89924 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:48,085 : INFO : EPOCH 1 - PROGRESS: at 92.55% examples, 89507 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:49,344 : INFO : EPOCH 1 - PROGRESS: at 93.42% examples, 88843 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:40:50,358 : INFO : EPOCH 1 - PROGRESS: at 94.20% examples, 88392 words/s, in_qsize 5, out_

2021-02-23 11:42:00,020 : INFO : EPOCH 2 - PROGRESS: at 42.38% examples, 49438 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:01,153 : INFO : EPOCH 2 - PROGRESS: at 42.75% examples, 49187 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:02,163 : INFO : EPOCH 2 - PROGRESS: at 43.43% examples, 49260 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:42:03,216 : INFO : EPOCH 2 - PROGRESS: at 44.61% examples, 49609 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:04,230 : INFO : EPOCH 2 - PROGRESS: at 45.59% examples, 49775 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:05,510 : INFO : EPOCH 2 - PROGRESS: at 46.04% examples, 49234 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:06,593 : INFO : EPOCH 2 - PROGRESS: at 46.55% examples, 49052 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:42:07,692 : INFO : EPOCH 2 - PROGRESS: at 47.06% examples, 48953 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:42:08,813 : INFO : EPOCH 2 - PROGRESS: at 47.44% examples, 48658 words/s, in_qsize 5, out_

2021-02-23 11:43:14,896 : INFO : EPOCH 3 - PROGRESS: at 34.14% examples, 109899 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:15,936 : INFO : EPOCH 3 - PROGRESS: at 35.97% examples, 109694 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:17,002 : INFO : EPOCH 3 - PROGRESS: at 36.80% examples, 107672 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:18,026 : INFO : EPOCH 3 - PROGRESS: at 37.55% examples, 106264 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:19,063 : INFO : EPOCH 3 - PROGRESS: at 38.63% examples, 105169 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:20,144 : INFO : EPOCH 3 - PROGRESS: at 39.88% examples, 104766 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:21,155 : INFO : EPOCH 3 - PROGRESS: at 41.28% examples, 104372 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:22,241 : INFO : EPOCH 3 - PROGRESS: at 42.32% examples, 103508 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:43:23,333 : INFO : EPOCH 3 - PROGRESS: at 43.53% examples, 103336 words/s, in_qsiz

2021-02-23 11:44:30,017 : INFO : EPOCH 4 - PROGRESS: at 22.48% examples, 86514 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:31,076 : INFO : EPOCH 4 - PROGRESS: at 23.66% examples, 86389 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:32,077 : INFO : EPOCH 4 - PROGRESS: at 24.95% examples, 86810 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:33,108 : INFO : EPOCH 4 - PROGRESS: at 26.27% examples, 87103 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:34,140 : INFO : EPOCH 4 - PROGRESS: at 27.30% examples, 87031 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:35,152 : INFO : EPOCH 4 - PROGRESS: at 28.79% examples, 87344 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:36,182 : INFO : EPOCH 4 - PROGRESS: at 30.07% examples, 87293 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:37,192 : INFO : EPOCH 4 - PROGRESS: at 30.95% examples, 86989 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:44:38,279 : INFO : EPOCH 4 - PROGRESS: at 32.44% examples, 87045 words/s, in_qsize 5, out_

2021-02-23 11:45:44,364 : INFO : EPOCH 5 - PROGRESS: at 11.32% examples, 86507 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:45,385 : INFO : EPOCH 5 - PROGRESS: at 12.98% examples, 87851 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:46,460 : INFO : EPOCH 5 - PROGRESS: at 14.57% examples, 88540 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:47,523 : INFO : EPOCH 5 - PROGRESS: at 16.24% examples, 89166 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:48,659 : INFO : EPOCH 5 - PROGRESS: at 17.70% examples, 89214 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:49,669 : INFO : EPOCH 5 - PROGRESS: at 18.84% examples, 89606 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:50,699 : INFO : EPOCH 5 - PROGRESS: at 20.60% examples, 90246 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:51,720 : INFO : EPOCH 5 - PROGRESS: at 21.96% examples, 90110 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:45:52,889 : INFO : EPOCH 5 - PROGRESS: at 23.38% examples, 89984 words/s, in_qsize 5, out_

2021-02-23 11:46:58,815 : INFO : training on a 53831415 raw words (36711561 effective words) took 447.2s, 82096 effective words/s
2021-02-23 11:46:58,907 : WARNING : consider setting layer size to a multiple of 4 for greater performance
2021-02-23 11:46:58,908 : INFO : collecting all words and their counts
2021-02-23 11:46:58,909 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2021-02-23 11:46:58,928 : INFO : PROGRESS: at sentence #10000, processed 112555 words, keeping 6354 word types
2021-02-23 11:46:58,950 : INFO : PROGRESS: at sentence #20000, processed 254983 words, keeping 10821 word types
2021-02-23 11:46:58,972 : INFO : PROGRESS: at sentence #30000, processed 367772 words, keeping 12635 word types
2021-02-23 11:46:59,004 : INFO : PROGRESS: at sentence #40000, processed 566573 words, keeping 17685 word types


70583
Threshold ferq = 0.00055 Percentage of correct answers: 68.75%


2021-02-23 11:46:59,031 : INFO : PROGRESS: at sentence #50000, processed 724294 words, keeping 20496 word types
2021-02-23 11:46:59,056 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:46:59,080 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:46:59,104 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:46:59,126 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:46:59,147 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:46:59,170 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:46:59,191 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:46:59,214 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keepin

2021-02-23 11:47:00,834 : INFO : Loading a fresh vocabulary
2021-02-23 11:47:00,944 : INFO : effective_min_count=1 retains 70583 unique words (100% of original 70583, drops 0)
2021-02-23 11:47:00,944 : INFO : effective_min_count=1 leaves 10766283 word corpus (100% of original 10766283, drops 0)
2021-02-23 11:47:01,142 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:47:01,146 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:47:01,147 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:47:01,354 : INFO : estimated required memory for 70583 words and 10 dimensions: 40938140 bytes
2021-02-23 11:47:01,355 : INFO : resetting layer weights
2021-02-23 11:47:16,419 : INFO : training model with 3 workers on 70583 vocabulary and 10 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:47:17,435 : INFO : EPOCH 1 - PROGRESS: at 9.15% examples, 666394 words/s, in_qsize 5, out_qsize 0
20

2021-02-23 11:48:11,566 : INFO : EPOCH 5 - PROGRESS: at 92.75% examples, 755779 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:48:12,180 : INFO : worker thread finished; awaiting finish of 2 more threads
2021-02-23 11:48:12,181 : INFO : worker thread finished; awaiting finish of 1 more threads
2021-02-23 11:48:12,188 : INFO : worker thread finished; awaiting finish of 0 more threads
2021-02-23 11:48:12,188 : INFO : EPOCH - 5 : training on 10766283 raw words (7340717 effective words) took 9.7s, 757449 effective words/s
2021-02-23 11:48:12,189 : INFO : training on a 53831415 raw words (36710807 effective words) took 55.8s, 658263 effective words/s
2021-02-23 11:48:12,235 : INFO : collecting all words and their counts
2021-02-23 11:48:12,236 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2021-02-23 11:48:12,253 : INFO : PROGRESS: at sentence #10000, processed 112555 words, keeping 6354 word types
2021-02-23 11:48:12,274 : INFO : PROGRESS: at sentence #20000, pr

70583
Threshold ferq = 0.00055 Percentage of correct answers: 57.5%


2021-02-23 11:48:12,390 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:48:12,411 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:48:12,432 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:48:12,450 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:48:12,473 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:48:12,492 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:48:12,514 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:48:12,539 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keeping 30297 word types
2021-02-23 11:48:12,559 : INFO : PROGRESS: at sentence #150000, processed 2044830 words, ke

2021-02-23 11:48:14,248 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:48:14,252 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:48:14,252 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:48:14,438 : INFO : estimated required memory for 70583 words and 100 dimensions: 91757900 bytes
2021-02-23 11:48:14,438 : INFO : resetting layer weights
2021-02-23 11:48:28,892 : INFO : training model with 3 workers on 70583 vocabulary and 100 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:48:29,934 : INFO : EPOCH 1 - PROGRESS: at 5.19% examples, 362382 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:48:30,934 : INFO : EPOCH 1 - PROGRESS: at 10.67% examples, 381311 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:48:31,966 : INFO : EPOCH 1 - PROGRESS: at 15.63% examples, 355556 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:48:32,967 : INFO : EPOCH 1 - PROGRESS: at 20.11% example

2021-02-23 11:49:28,786 : INFO : EPOCH 4 - PROGRESS: at 71.01% examples, 468059 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:49:29,800 : INFO : EPOCH 4 - PROGRESS: at 75.68% examples, 461644 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:49:30,806 : INFO : EPOCH 4 - PROGRESS: at 82.26% examples, 461034 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:49:31,816 : INFO : EPOCH 4 - PROGRESS: at 87.30% examples, 458414 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:49:32,846 : INFO : EPOCH 4 - PROGRESS: at 93.13% examples, 453195 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:49:33,854 : INFO : EPOCH 4 - PROGRESS: at 99.59% examples, 451438 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:49:33,893 : INFO : worker thread finished; awaiting finish of 2 more threads
2021-02-23 11:49:33,899 : INFO : worker thread finished; awaiting finish of 1 more threads
2021-02-23 11:49:33,904 : INFO : worker thread finished; awaiting finish of 0 more threads
2021-02-23 11:49:33,904 : INFO : EPOCH - 4 : trai

70583
Threshold ferq = 0.00055 Percentage of correct answers: 65.0%


2021-02-23 11:49:49,495 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:49:49,522 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:49:49,544 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:49:49,569 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:49:49,588 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:49:49,613 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:49:49,632 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:49:49,657 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:49:49,685 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keep

2021-02-23 11:49:51,392 : INFO : effective_min_count=1 leaves 10766283 word corpus (100% of original 10766283, drops 0)
2021-02-23 11:49:51,594 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:49:51,599 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:49:51,600 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:49:51,811 : INFO : estimated required memory for 70583 words and 1000 dimensions: 599955500 bytes
2021-02-23 11:49:51,812 : INFO : resetting layer weights
2021-02-23 11:50:06,578 : INFO : training model with 3 workers on 70583 vocabulary and 1000 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:50:07,597 : INFO : EPOCH 1 - PROGRESS: at 1.52% examples, 87194 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:50:08,676 : INFO : EPOCH 1 - PROGRESS: at 2.91% examples, 91193 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:50:09,722 : INFO : EPOCH 1 - PROGRESS: at 4.43% e

2021-02-23 11:51:18,363 : INFO : EPOCH 1 - PROGRESS: at 83.49% examples, 85940 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:51:19,460 : INFO : EPOCH 1 - PROGRESS: at 84.94% examples, 86037 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:20,466 : INFO : EPOCH 1 - PROGRESS: at 85.92% examples, 86064 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:21,613 : INFO : EPOCH 1 - PROGRESS: at 86.96% examples, 86030 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:22,642 : INFO : EPOCH 1 - PROGRESS: at 88.24% examples, 86127 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:23,649 : INFO : EPOCH 1 - PROGRESS: at 89.45% examples, 86151 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:24,765 : INFO : EPOCH 1 - PROGRESS: at 90.95% examples, 86216 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:25,797 : INFO : EPOCH 1 - PROGRESS: at 92.55% examples, 86371 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:51:26,912 : INFO : EPOCH 1 - PROGRESS: at 94.01% examples, 86440 words/s, in_qsize 5, out_

2021-02-23 11:52:34,023 : INFO : EPOCH 2 - PROGRESS: at 76.24% examples, 89621 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:35,036 : INFO : EPOCH 2 - PROGRESS: at 77.98% examples, 89811 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:36,135 : INFO : EPOCH 2 - PROGRESS: at 79.22% examples, 89749 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:37,204 : INFO : EPOCH 2 - PROGRESS: at 80.61% examples, 89846 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:38,284 : INFO : EPOCH 2 - PROGRESS: at 81.96% examples, 89908 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:39,342 : INFO : EPOCH 2 - PROGRESS: at 82.83% examples, 89610 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:40,382 : INFO : EPOCH 2 - PROGRESS: at 83.76% examples, 89432 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:41,410 : INFO : EPOCH 2 - PROGRESS: at 85.19% examples, 89377 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:52:42,412 : INFO : EPOCH 2 - PROGRESS: at 85.85% examples, 89072 words/s, in_qsize 5, out_

2021-02-23 11:53:47,715 : INFO : EPOCH 3 - PROGRESS: at 82.77% examples, 109531 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:48,774 : INFO : EPOCH 3 - PROGRESS: at 84.19% examples, 109512 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:49,863 : INFO : EPOCH 3 - PROGRESS: at 85.58% examples, 109206 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:50,878 : INFO : EPOCH 3 - PROGRESS: at 86.62% examples, 109073 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:51,903 : INFO : EPOCH 3 - PROGRESS: at 88.04% examples, 109049 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:52,951 : INFO : EPOCH 3 - PROGRESS: at 89.67% examples, 109071 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:53,977 : INFO : EPOCH 3 - PROGRESS: at 91.47% examples, 109225 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:55,011 : INFO : EPOCH 3 - PROGRESS: at 93.21% examples, 109256 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:53:56,086 : INFO : EPOCH 3 - PROGRESS: at 94.83% examples, 109237 words/s, in_qsiz

2021-02-23 11:55:00,979 : INFO : EPOCH 4 - PROGRESS: at 91.24% examples, 109020 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:55:02,022 : INFO : EPOCH 4 - PROGRESS: at 93.13% examples, 109148 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:55:03,165 : INFO : EPOCH 4 - PROGRESS: at 94.83% examples, 109121 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:55:04,218 : INFO : EPOCH 4 - PROGRESS: at 96.70% examples, 109221 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:55:05,275 : INFO : EPOCH 4 - PROGRESS: at 98.71% examples, 109329 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:55:05,974 : INFO : worker thread finished; awaiting finish of 2 more threads
2021-02-23 11:55:06,003 : INFO : worker thread finished; awaiting finish of 1 more threads
2021-02-23 11:55:06,020 : INFO : worker thread finished; awaiting finish of 0 more threads
2021-02-23 11:55:06,021 : INFO : EPOCH - 4 : training on 10766283 raw words (7342811 effective words) took 67.1s, 109493 effective words/s
2021-02-23 11:55:07,071

2021-02-23 11:56:12,973 : INFO : worker thread finished; awaiting finish of 0 more threads
2021-02-23 11:56:12,974 : INFO : EPOCH - 5 : training on 10766283 raw words (7342172 effective words) took 66.9s, 109672 effective words/s
2021-02-23 11:56:12,975 : INFO : training on a 53831415 raw words (36711423 effective words) took 366.4s, 100196 effective words/s
2021-02-23 11:56:13,061 : WARNING : consider setting layer size to a multiple of 4 for greater performance
2021-02-23 11:56:13,063 : INFO : collecting all words and their counts
2021-02-23 11:56:13,063 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2021-02-23 11:56:13,080 : INFO : PROGRESS: at sentence #10000, processed 112555 words, keeping 6354 word types
2021-02-23 11:56:13,101 : INFO : PROGRESS: at sentence #20000, processed 254983 words, keeping 10821 word types
2021-02-23 11:56:13,118 : INFO : PROGRESS: at sentence #30000, processed 367772 words, keeping 12635 word types
2021-02-23 11:56:13,149 : I

70583
Threshold ferq = 0.00055 Percentage of correct answers: 72.5%


2021-02-23 11:56:13,199 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:56:13,222 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:56:13,245 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:56:13,268 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:56:13,285 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:56:13,307 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:56:13,326 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:56:13,347 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:56:13,371 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keep

2021-02-23 11:56:15,664 : INFO : effective_min_count=1 leaves 10766283 word corpus (100% of original 10766283, drops 0)
2021-02-23 11:56:15,841 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:56:15,844 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:56:15,845 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:56:16,024 : INFO : estimated required memory for 70583 words and 10 dimensions: 40938140 bytes
2021-02-23 11:56:16,025 : INFO : resetting layer weights
2021-02-23 11:56:28,728 : INFO : training model with 3 workers on 70583 vocabulary and 10 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:56:29,745 : INFO : EPOCH 1 - PROGRESS: at 10.04% examples, 726647 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:56:30,749 : INFO : EPOCH 1 - PROGRESS: at 22.18% examples, 762261 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:56:31,765 : INFO : EPOCH 1 - PROGRESS: at 32.63% e

2021-02-23 11:57:16,741 : INFO : PROGRESS: at sentence #20000, processed 254983 words, keeping 10821 word types
2021-02-23 11:57:16,758 : INFO : PROGRESS: at sentence #30000, processed 367772 words, keeping 12635 word types
2021-02-23 11:57:16,787 : INFO : PROGRESS: at sentence #40000, processed 566573 words, keeping 17685 word types
2021-02-23 11:57:16,811 : INFO : PROGRESS: at sentence #50000, processed 724294 words, keeping 20496 word types
2021-02-23 11:57:16,835 : INFO : PROGRESS: at sentence #60000, processed 869309 words, keeping 22253 word types
2021-02-23 11:57:16,857 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types


70583
Threshold ferq = 0.00055 Percentage of correct answers: 52.5%


2021-02-23 11:57:16,877 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:57:16,899 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:57:16,918 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:57:16,938 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:57:16,957 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:57:16,978 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:57:17,004 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keeping 30297 word types
2021-02-23 11:57:17,024 : INFO : PROGRESS: at sentence #150000, processed 2044830 words, keeping 31192 word types
2021-02-23 11:57:17,042 : INFO : PROGRESS: at sentence #160000, processed 2160734 words, k

2021-02-23 11:57:18,692 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:57:18,693 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:57:18,874 : INFO : estimated required memory for 70583 words and 100 dimensions: 91757900 bytes
2021-02-23 11:57:18,875 : INFO : resetting layer weights
2021-02-23 11:57:30,850 : INFO : training model with 3 workers on 70583 vocabulary and 100 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:57:31,877 : INFO : EPOCH 1 - PROGRESS: at 7.49% examples, 546491 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:57:32,879 : INFO : EPOCH 1 - PROGRESS: at 16.44% examples, 565608 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:57:33,883 : INFO : EPOCH 1 - PROGRESS: at 24.62% examples, 565406 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:57:34,909 : INFO : EPOCH 1 - PROGRESS: at 32.44% examples, 563251 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:57:35,919 : INFO : EPOCH 1

2021-02-23 11:58:27,097 : INFO : EPOCH 5 - PROGRESS: at 16.34% examples, 559447 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:28,110 : INFO : EPOCH 5 - PROGRESS: at 24.53% examples, 559613 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:29,125 : INFO : EPOCH 5 - PROGRESS: at 32.19% examples, 557263 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:30,134 : INFO : EPOCH 5 - PROGRESS: at 39.33% examples, 555378 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:31,134 : INFO : EPOCH 5 - PROGRESS: at 46.62% examples, 553332 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:32,145 : INFO : EPOCH 5 - PROGRESS: at 54.55% examples, 554749 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:33,148 : INFO : EPOCH 5 - PROGRESS: at 61.28% examples, 552989 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:34,160 : INFO : EPOCH 5 - PROGRESS: at 68.65% examples, 552723 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:35,172 : INFO : EPOCH 5 - PROGRESS: at 75.31% examples, 551733 words/s, in_qsiz

70583
Threshold ferq = 0.00055 Percentage of correct answers: 68.75%


2021-02-23 11:58:38,582 : INFO : PROGRESS: at sentence #70000, processed 1006992 words, keeping 23483 word types
2021-02-23 11:58:38,602 : INFO : PROGRESS: at sentence #80000, processed 1138953 words, keeping 24604 word types
2021-02-23 11:58:38,625 : INFO : PROGRESS: at sentence #90000, processed 1271320 words, keeping 25783 word types
2021-02-23 11:58:38,644 : INFO : PROGRESS: at sentence #100000, processed 1381760 words, keeping 26648 word types
2021-02-23 11:58:38,667 : INFO : PROGRESS: at sentence #110000, processed 1513702 words, keeping 27511 word types
2021-02-23 11:58:38,687 : INFO : PROGRESS: at sentence #120000, processed 1631919 words, keeping 28085 word types
2021-02-23 11:58:38,712 : INFO : PROGRESS: at sentence #130000, processed 1765792 words, keeping 28818 word types
2021-02-23 11:58:38,736 : INFO : PROGRESS: at sentence #140000, processed 1923924 words, keeping 30297 word types
2021-02-23 11:58:38,757 : INFO : PROGRESS: at sentence #150000, processed 2044830 words, ke

2021-02-23 11:58:40,518 : INFO : deleting the raw counts dictionary of 70583 items
2021-02-23 11:58:40,521 : INFO : sample=0.00055 downsamples 87 most-common words
2021-02-23 11:58:40,522 : INFO : downsampling leaves estimated 7342202 word corpus (68.2% of prior 10766283)
2021-02-23 11:58:40,717 : INFO : estimated required memory for 70583 words and 1000 dimensions: 599955500 bytes
2021-02-23 11:58:40,719 : INFO : resetting layer weights
2021-02-23 11:58:53,458 : INFO : training model with 3 workers on 70583 vocabulary and 1000 features, using sg=1 hs=0 sample=0.00055 negative=5 window=5
2021-02-23 11:58:54,530 : INFO : EPOCH 1 - PROGRESS: at 1.80% examples, 102129 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:55,581 : INFO : EPOCH 1 - PROGRESS: at 3.63% examples, 109159 words/s, in_qsize 5, out_qsize 0
2021-02-23 11:58:56,615 : INFO : EPOCH 1 - PROGRESS: at 4.84% examples, 107931 words/s, in_qsize 6, out_qsize 0
2021-02-23 11:58:57,741 : INFO : EPOCH 1 - PROGRESS: at 6.18% example

2021-02-23 12:00:02,320 : INFO : EPOCH 2 - PROGRESS: at 1.80% examples, 105298 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:03,330 : INFO : EPOCH 2 - PROGRESS: at 3.49% examples, 109799 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:04,437 : INFO : EPOCH 2 - PROGRESS: at 4.78% examples, 105887 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:05,467 : INFO : EPOCH 2 - PROGRESS: at 6.11% examples, 107364 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:06,547 : INFO : EPOCH 2 - PROGRESS: at 7.49% examples, 106058 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:07,621 : INFO : EPOCH 2 - PROGRESS: at 9.23% examples, 107259 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:08,625 : INFO : EPOCH 2 - PROGRESS: at 10.97% examples, 108462 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:09,639 : INFO : EPOCH 2 - PROGRESS: at 12.73% examples, 109235 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:00:10,665 : INFO : EPOCH 2 - PROGRESS: at 14.57% examples, 109654 words/s, in_qsize 5, o

2021-02-23 12:01:14,863 : INFO : EPOCH 3 - PROGRESS: at 7.51% examples, 107000 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:15,946 : INFO : EPOCH 3 - PROGRESS: at 9.26% examples, 107940 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:17,046 : INFO : EPOCH 3 - PROGRESS: at 11.04% examples, 108581 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:18,111 : INFO : EPOCH 3 - PROGRESS: at 12.98% examples, 109500 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:19,203 : INFO : EPOCH 3 - PROGRESS: at 14.95% examples, 109826 words/s, in_qsize 4, out_qsize 1
2021-02-23 12:01:20,311 : INFO : EPOCH 3 - PROGRESS: at 16.82% examples, 109879 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:21,337 : INFO : EPOCH 3 - PROGRESS: at 18.22% examples, 110184 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:22,347 : INFO : EPOCH 3 - PROGRESS: at 19.99% examples, 110059 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:01:23,431 : INFO : EPOCH 3 - PROGRESS: at 21.96% examples, 110381 words/s, in_qsize 

2021-02-23 12:02:27,906 : INFO : EPOCH 4 - PROGRESS: at 16.82% examples, 110250 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:28,910 : INFO : EPOCH 4 - PROGRESS: at 18.17% examples, 110159 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:29,965 : INFO : EPOCH 4 - PROGRESS: at 19.99% examples, 110147 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:31,044 : INFO : EPOCH 4 - PROGRESS: at 21.96% examples, 110485 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:32,162 : INFO : EPOCH 4 - PROGRESS: at 23.66% examples, 110462 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:33,254 : INFO : EPOCH 4 - PROGRESS: at 25.28% examples, 110571 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:34,269 : INFO : EPOCH 4 - PROGRESS: at 26.64% examples, 110384 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:35,299 : INFO : EPOCH 4 - PROGRESS: at 28.33% examples, 110472 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:02:36,303 : INFO : EPOCH 4 - PROGRESS: at 30.07% examples, 110740 words/s, in_qsiz

2021-02-23 12:03:40,781 : INFO : EPOCH 5 - PROGRESS: at 23.30% examples, 110694 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:41,783 : INFO : EPOCH 5 - PROGRESS: at 24.71% examples, 110101 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:42,823 : INFO : EPOCH 5 - PROGRESS: at 25.88% examples, 108960 words/s, in_qsize 6, out_qsize 0
2021-02-23 12:03:43,829 : INFO : EPOCH 5 - PROGRESS: at 27.30% examples, 108928 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:44,878 : INFO : EPOCH 5 - PROGRESS: at 29.08% examples, 108983 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:45,891 : INFO : EPOCH 5 - PROGRESS: at 30.60% examples, 109251 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:46,955 : INFO : EPOCH 5 - PROGRESS: at 32.19% examples, 108898 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:47,970 : INFO : EPOCH 5 - PROGRESS: at 33.20% examples, 108549 words/s, in_qsize 5, out_qsize 0
2021-02-23 12:03:49,012 : INFO : EPOCH 5 - PROGRESS: at 35.25% examples, 108992 words/s, in_qsiz

70583
Threshold ferq = 0.00055 Percentage of correct answers: 71.25%
Dimension 10 mean accuracy =  54.583333333333336
Dimension 100 mean accuracy =  68.33333333333333
Dimension 1000 mean accuracy =  70.83333333333333


## Accuracy with different dimension (Word2Vec)
Dimension 10 mean accuracy =  54.583333333333336

Dimension 100 mean accuracy =  68.33333333333333

Dimension 1000 mean accuracy =  70.83333333333333

In [11]:
import numpy as np
import text_functions as tf
import nltk

#@author: The first version of this code is the courtesy of Vadim Selyanik

threshold = 15000 # Frequency threshold in the corpus ??
dimension = 2000 # Dimensionality for high-dimensional vectors
lemmatizer = nltk.WordNetLemmatizer()  # create an instance of lemmatizer
ones_number = 2 # number of nonzero elements in randomly generated high-dimensional vectors
window_size = 2 #number of neighboring words to consider both back and forth. In other words number of words before/after current word
zero_vector = np.zeros(dimension)
test_name = "new_toefl.txt" # file with TOEFL dataset
data_file_name = "lemmatized.text" # file with the text corpus

amount_dictionary = {}

# Count how many times each word appears in the corpus
text_file = open(data_file_name, "r")
for line in text_file:
    if line != "\n":
        words = line.split()
        for word in words:
            if amount_dictionary.get(word) is None:
                amount_dictionary[word] = 1
            else:
                amount_dictionary[word] += 1
text_file.close()

dictionary = {} #vocabulary and corresponing random high-dimensional vectors
word_space = {} #embedings




#Create a dictionary with the assigned random high-dimensional vectors
text_file = open(data_file_name, "r")
for line in text_file: #read line in the file
    words = line.split() # extract words from the line
    for word in words:  # for each word
        if dictionary.get(word) is None: # If the word was not yed added to the vocabulary
            if amount_dictionary[word] < threshold:
                dictionary[word] = tf.get_random_word_vector(dimension, ones_number) # assign a  
            else:
                dictionary[word] = np.zeros(dimension) # frequent words are assigned with empty vectors. In a way they will not contribute to the word embedding

text_file.close()
#Note that in order to save time we only create embeddings for the words needed in the TOEFL task

#Find all unique words amongst TOEFL tasks and initialize their embeddings to zeros    
number_of_tests = 0
text_file = open(test_name, "r") #open TOEFL tasks
for line in text_file:
        words = line.split()
        words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                 words] # lemmatize words in the current test
        word_space[words[0]] = np.zeros(dimension)
        word_space[words[1]] = np.zeros(dimension)
        word_space[words[2]] = np.zeros(dimension)
        word_space[words[3]] = np.zeros(dimension)
        word_space[words[4]] = np.zeros(dimension)
        number_of_tests += 1
text_file.close()


ModuleNotFoundError: No module named 'text_functions'

In [ ]:
# Processing the text to build the embeddings 
text_file = open(data_file_name, "r")
lines = [[],[],[],[]] # neighboring lines
i = 2
while i < 4:
        line = "\n"
        while line == "\n":
            line = text_file.readline()
        lines[i] = line.split()
        i += 1

line = text_file.readline()
while line != "":
        if line != "\n":
            lines.append(line.split())
            words = lines[2]
            length = len(words)
            i = 0
            while i < length:
                if not (word_space.get(words[i]) is None):
                    k = 1
                    word_space_vector = word_space[words[i]]
                    while (i - k >= 0) and (k <= window_size): #process left neighbors of the focus word
                        word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[words[i - k]], -1))         
                        k += 1
                    # Handle different situations if there was not enough neighbors on the left in the current line    
                    if k <= window_size and (len(lines[1])>0): 
                        if len(lines[1]) < 2:
                            if k != 1: #if one word on the left was already added
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[1][0]], -1)) #update word embedding
                            else:
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                              np.roll(dictionary[lines[1][0]], -1)) #update word embedding
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                              np.roll(dictionary[lines[0][len(lines[0]) - 1]], -1)) #update word embedding
                        else:
                            if k != 1:
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                              np.roll(dictionary[lines[1][len(lines[1]) - 1]], -1)) #update word embedding
                            else:
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                              np.roll(dictionary[lines[1][len(lines[1]) - 1]], -1)) #update word embedding
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                              np.roll(dictionary[lines[1][len(lines[1]) - 2]], -1)) #update word embedding

                    k = 1
                    while (i + k < length) and (k <= window_size): #process right neighbors of the focus word
                        word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[words[i + k]], 1)) #update word embedding
                        k += 1
                    if k <= window_size:
                        if len(lines[3]) < 2:
                            if k != 1:
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[3][0]], 1)) #update word embedding
                            else:
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[3][0]], 1)) #update word embedding
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[4][0]], 1)) #update word embedding
                        else:
                            if k != 1:
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[3][0]], 1)) #update word embedding
                            else:
                                word_space[words[i]] = np.add(word_space[words[i]], np.roll(dictionary[lines[3][0]], 1)) #update word embedding
                                word_space[words[i]] = np.add(word_space[words[i]],
                                                          np.roll(dictionary[lines[3][1]], 1))

                i += 1
            lines.pop(0)
        line = text_file.readline()



#Testing of the embeddings on TOEFL
a = 0.0 # accuracy of the encodings    
i = 0
text_file = open(test_name, 'r')
right_answers = 0.0 # variable for correct answers
number_skipped_tests = 0.0 # some tests could be skipped if there are no corresponding words in the vocabulary extracted from the training corpus
while i < number_of_tests:
        line = text_file.readline() #read line in the file
        words = line.split()  # extract words from the line
        words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                  words]  # lemmatize words in the current test
        try:
            
            if not(amount_dictionary.get(words[0]) is None): # check if there word in the corpus for the query word
                k = 1
                while k < 5:
                    # if amount_dictionary.get(words[k]) is None:
                    #     word_space[words[k]] = np.random.randn(dimension)
                    if np.array_equal(word_space[words[k]], zero_vector): # if no representation was learnt assign a random vector
                        word_space[words[k]] = np.random.randn(dimension)
                    k += 1
                right_answers += tf.get_answer_mod([word_space[words[0]],word_space[words[1]],word_space[words[2]],
                            word_space[words[3]],word_space[words[4]]]) #check if word is predicted right
        except KeyError: # if there is no representation for the query vector than skip
            number_skipped_tests += 1
            print("skipped test: " + str(i) + "; Line: " + str(words))
        except IndexError:
            print(i)
            print(line)
            print(words)
            break
        i += 1
text_file.close()
a += 100 * right_answers / number_of_tests
print(str(dimension) + " Percentage of correct answers: " + str(100 * right_answers / number_of_tests) + "%")



